sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/errorhandling/BadRecordRouter.java

```diff
@@ -0,0 +1,129 @@
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.Serializable;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.transforms.DoFn.MultiOutputReceiver;
+import org.apache.beam.sdk.transforms.errorhandling.BadRecord.Failure;
+import org.apache.beam.sdk.transforms.errorhandling.BadRecord.Record;
+import org.apache.beam.sdk.util.Preconditions;
+import org.apache.beam.sdk.values.TupleTag;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public interface BadRecordRouter extends Serializable {
+
+  BadRecordRouter THROWING_ROUTER = new ThrowingBadRecordRouter();
+
+  BadRecordRouter RECORDING_ROUTER = new RecordingBadRecordRouter();
+
+  TupleTag<BadRecord> BAD_RECORD_TAG = new TupleTag<>();
+
+  <RecordT> void route(
+      MultiOutputReceiver outputReceiver,
+      RecordT record,
+      @Nullable Coder<RecordT> coder,
+      @Nullable Exception exception,
+      String description,
+      String failingTransform)
+      throws Exception;
+
+  class ThrowingBadRecordRouter implements BadRecordRouter {
+
+    @Override
+    public <RecordT> void route(
+        MultiOutputReceiver outputReceiver,
+        RecordT record,
+        @Nullable Coder<RecordT> coder,
+        @Nullable Exception exception,
+        String description,
+        String failingTransform)
+        throws Exception {
+      if (exception != null) {
+        throw exception;
+      }
+    }
+  }
+
+  class RecordingBadRecordRouter implements BadRecordRouter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(RecordingBadRecordRouter.class);
+
+    @Override
+    public <RecordT> void route(
+        MultiOutputReceiver outputReceiver,
+        RecordT record,
+        @Nullable Coder<RecordT> coder,
+        @Nullable Exception exception,
+        String description,
+        String failingTransform)
+        throws Exception {
+      Preconditions.checkArgumentNotNull(record);
+      ObjectWriter objectWriter = new ObjectMapper().writer().withDefaultPrettyPrinter();
+
+      // Build up record information
+      BadRecord.Record.Builder recordBuilder = Record.builder();
+      try {
+        recordBuilder.setJsonRecord(objectWriter.writeValueAsString(record));
+      } catch (Exception e) {
+        LOG.error("Unable to serialize record as JSON. Human readable record will be null", e);
+      }
+
+      // We will sometimes not have a coder for a failing record, for example if it has already been
+      // modified within the dofn.
+      if (coder != null) {
+        recordBuilder.setCoder(coder.toString());
+
+        try {
+          ByteArrayOutputStream stream = new ByteArrayOutputStream();
+          coder.encode(record, stream);
+          byte[] bytes = stream.toByteArray();
+          recordBuilder.setEncodedRecord(bytes);
```

Comment:
So a common case we will want to handle is malformed input bytes when trying to read/parse a record. How is that handled here?